README.md
# Development Environment

This setup relies on Vagrant and VirtualBox (or [VMWare Fusion](#vmware-provider) if that's what you prefer) running on Mac OS X to power the virtualized developer environment. These dependencies are installed as part of the setup process using [Homebrew](http://brew.sh) and [Homebrew Cask](https://caskroom.github.io/).

It is setup with different machines running different versions of PHP (PHP 7.0 is the default machine). The machines provide all the necessary components to build on Magento 1 and Magento 2, including support for running multiple PHP / MySql versions side-by-side if necessary ([see below for details](#virtual-machines)). The nodes run a traditional LAMP stack, with Nginx sitting in front of Apache as a proxy for static assets and for SSL termination. It also includes [Xdebug](http://xdebug.org) pre-configured to connect to your IDE on the host machine.

## System Requirements

* Mac OS X 10.11 or later
* An HFS+ or APFS **Case-sensitive** partition mounted at `/Volumes/Server` or `/server`

    *Note: The environment should install and run from a case-insensitive mount, but this is not recommended for two reasons: a) the majority of deployments are done to case-sensitive file-systems, so development done on a case-sensitive mount is less error prone (ex: autoloaders may find a class in development, then fail on production); b) mysql will behave differently as it pertains to [identifier case sensitivity](https://dev.mysql.com/doc/refman/5.0/en/identifier-case-sensitivity.html) potentially causing unexpected behavior*

## Environment Setup

1. The install process will install brew on the host machine for gathering dependencies where not already present. If you already have brew installed, however, it is recommended to run the following commands, then cleanup any major issues it reports:

    ```bash
    brew update
    brew doctor
    ```

2. Install technical dependencies and setup the environment, entering your account password when prompted (this may happen a few times):

    ```bash
    curl -s https://raw.githubusercontent.com/classyllama/devenv/master/vagrant/bin/install.sh | bash
    source /etc/profile
    ```

3. Run the following to start up the virtual machines. This may take a while on first run

    ```bash
    cd /server
    vagrant up
    ```

4. To SSH into the vm, you can use `vcd` or `vcd web` to connect and automatically mirror your working directory

### Optional Steps

1. Install the compass tools used for scss compilation

    ```bash
    sudo gem update --system
    sudo sudo gem install -n /usr/local/bin
    ```
      
2. Generate an RSA key pair. The generated public key will be used to authenticate remote SSH connections

    ```bash
    ssh-keygen -f ~/.ssh/id_rsa
    ```

    *Note: When prompted, enter a memorable passphrase (you’ll need to use it later)*

3. Because of GitHub's rate limits on their API it can happen that Composer will silently fail when running the m2setup.sh tool. To prevent this from happening, create an OAuth token via the [GitHub Settings](https://github.com/settings/tokens) area in your GitHub account. You can read more about these tokens [here](https://github.com/blog/1509-personal-api-tokens). Add this token to the composer configuration by running:

    ```bash
    composer config -g github-oauth.github.com "<oauthtoken>"
    ```

4. Create a Magento 2 build available at m2.demo:

    ```bash
    vagrant ssh -- m2setup.sh --sampledata --hostname=m2.demo
    echo "10.19.89.15 m2.demo" | sudo tee -a /etc/hosts > /dev/null
    ```

### Quick Reference

| hostname      | ip           | role     | autostart | description                                        |
| ------------- | ------------ | -------- | --------- | -------------------------------------------------- |
| dev-host      | 10.19.89.1   | host     | n/a       | this is the host machine for the environment       |
| [dev-web71]   | 10.19.89.15  | app      | no        | App node running PHP 7.1 / Percona Server 5.6      |
| [dev-web70]   | 10.19.89.14  | app      | **yes**   | App node running PHP 7.0 / Percona Server 5.6      |
| [dev-web56]   | 10.19.89.10  | app      | no        | App node running PHP 5.6 / Percona Server 5.6      |
| [dev-web55]   | 10.19.89.11  | app      | no        | App node running PHP 5.5 / Percona Server 5.6      |

## Virtual Machines

### Web Application
This node is setup to run services required to run web applications. Nginx is setup to deliver static assets directly and act as a proxy for anything else. Apache is setup with mod_php to delivery the web application and sits behind Nginx on an internal port. Redis has been setup for a cache data store such that it never writes information to disk.

Run `vhosts.sh` to generate vhosts for all sites and reload apache. This will be automatically run once when the machine is provisioned, and may be subsequently run within the guest environment (use --help for available options).

The IP address of this node is fixed at `10.19.89.10`. This IP should be used in `/etc/hosts` on the host machine to facilitate loading applications running within the vm from a browser on the host.

#### Virtual Host Configuration
Virtual hosts are created automatically for each site by running the `vhosts.sh` script. These .conf files are based on a template, or may manually be configured on a per-site basis by placing a `.<service>.conf` file in the root site directory where `<service>` is the name of the service the file is to configure (such as nginx or httpd).
    
To configure the virtual host configuration and reload services, run `vhosts.sh` within the guest machine. Running `vhosts.sh --reset-config --reset-certs` will wipe out all generated certificates and service configuration, creating it from scratch.

The `vhosts.sh` script looks for the pretense of three locations within each directory contained by `/sites` to determine if a given directory found in `/sites` is in fact in need of a virtual host. These locations are as follows:

* /sites/example.dev/pub
* /sites/example.dev/html
* /sites/example.dev/htdocs

If any of these three paths exist, a virtual host will be created based on the template found in `/server/vagrant/etc/httpd/sites.d/__vhost.conf.template`. The `DocumentRoot` will be configured using the first of the above three paths found for a given site directory. The `ServerName` will match the name of the sites directory (example.dev above) and a wildcard `ServerAlias` is included to support relevant sub-domains. When a file is found at `/sites/example.dev/.vhost.conf` it will be used in leu of the template file. Any updates to this file will be applied to the host configuration on subsequent runs of the `vhosts.sh` script.

#### PHP Versions

The default web70 node has PHP 7.0.x from the IUS repository. PHP 5.6 is also available via the web56 node, but it will not start automatically. To use it, start via `vagrant up web56`, then configure your local hosts file to point the site needing this specific version of PHP to the correct machine instance.

#### SSL
When the `web` VM is provisioned, a root CA is automatically generated and stored at `/server/.shared/ssl/rootca/certs/ca.cert.pem` if it does not already exist.
During vhost discovery and configuration, a wildcard cert, signed by the root CA, is automatically generated for it. Nginx is configured accordingly.

This means that all vhosts support SSL on both the naked domain and any immediate subdomain. 
Since these certs are all signed by the persistent root CA, if the root CA is added to the host as a trusted cert, the SSL cert for any vhost will automatically be valid.

##### Mac
To add the generated root CA to your trusted certs list on the host machine, run this command (after vagrant up has been run):

```bash
sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain /server/.shared/ssl/rootca/certs/ca.cert.pem
```

For Firefox you will need to add the certificate authority manually through Firefox's interface.
* Export the Vagrant DevEnv certificate authority from your System Keychain (right-click, 'Export "Vagrant DevEnv"')
* Navigate to your Firefox preferences
  * Click "Privacy & Security"
  * Scroll down to "Certificates" and click "View Certificates"
  * Select the "Authorities" section and click "Import..."
  * Select the Vagrant DevEnv cert authority exported in the first step


##### Windows
To add the generated root CA to your certificate manager on Windows you will need to copy the `ca.cert.pem` file to a location on your Windows system like `C:\certs\ca.cert.pem` and then open a Command Prompt window in Administrator mode to execute the following command

```bash
certutil –addstore -enterprise –f "Root" c:\certs\ca.cert.pem
```

### Database Server

Each machine has an instance of Percona Server 5.6.x installed. Since this is a development environment, the root mysql password has been left blank. The data from the default machine is persisted in `/server/mysql/data`, data from other nodes persisted at `/server/mysql/<NAME>` where `<NAME>` is the name of machine the data is for.

To allow for custom database settings without modifying the default my.cnf file directly, files found at `vagrant/etc/my.cnf.d/*.cnf` will be copied onto this node and are applied via the `!includedir` directive in the `/etc/my.cnf` defaults file. Example use case: create the file `vagrant/etc/my.cnf.d/lower_case_table_names.cnf` with the following contents and then re-provision the vm:

    [mysqld]
    lower_case_table_names = 1

***WARNING:*** Because data is persisted to the host machine via an NFS mount, attempts to forcefully shutdown (aka run `vagrant destroy`) a machine may cause data corruption and will fail subsequent mysql start operations unless the vm has first been halted and/or the mysql service stopped gracefully prior to vm destruction. The recommended sequence to wipe the vm and create from scratch is halt, destroy, then up.

#### Common Problems
##### Percona Server fails to start
When this happens you'll see something like the following when attempting to provision or boot the vm:

    TASK [Starting mysql service] **************************************************
    fatal: [web]: FAILED! => {"changed": false, "failed": true, "msg": "Starting MySQL (Percona Server). ERROR! The server quit without updating PID file (/var/lib/mysql/dev-web70.pid).\n"}

This happens (per above warning) when the mysql service fails to shutdown cleanly. To solve this issue, proceed through the following steps:

***WARNING:*** If this is done and there is a running mysql instance using these ib* files, irreversible data corruption could occur. Please be careful!

1. Verify that Virtual Box reports none of the virtual machines as still running. If machines are still running, you will need to halt each of them before proceeding.

    ```bash
    VBoxManage list runningvms | grep "Server_"
    ```

2. Restart the `rpc.lockd` service on the host

    ```bash
    sudo launchctl unload /System/Library/LaunchDaemons/com.apple.lockd.plist
    sudo launchctl load /System/Library/LaunchDaemons/com.apple.lockd.plist
    ```

    Starting in MacOS 10.12.4 the rpc.lockd service is protected by System Integrity Protection, preventing you from reloading it via launchctl. If you see the message `Operation not permitted while System Integrity Protection is engaged` you will need to kill the `rpc.lockd` service instead. The service configuration has KeepAlive enabled, so the desired effect is accomplised, although in a much more heavy-handed fashion. Do this by running `sudo kill <PID>` where `<PID>` is replaced with the number `lsof` (per following step) displays under the PID column for the `rpc.lockd` command.

3. Verify no locks exist on your `ib*` files (command should return nothing)

    ```bash
    sudo lsof /server/mysql/*/ib*
    ```

4. Destroy and restart the virtual machine

    ```bash
    vagrant destroy -f
    vagrant up
    ```

If the above does not succeed in bringing it back online, try rebooting the host machine. If that still does not solve the issue, it is likely you will have to help mysqld out a bit with recovery. Check `/var/log/mysqld.log` for more info.

##### Virtual Box Cask
 if you get a error this error while running `curl -s https://raw.githubusercontent.com/classyllama/devenv/master/vagrant/bin/install.sh  https://raw.githubusercontent.com/classyllama/devenv/master/vagrant/bin/devbox.sh | bash /dev/stdin --lib-mode` 
 
 ```Error: Failure while executing; /usr/bin/sudo -E -- env LOGNAME=joshvahl USER=joshvahl USERNAME=joshvahl /usr/sbin/installer -pkg /usr/local/Caskroom/virtualbox/6.0.10,132072/VirtualBox.pkg -target /` exited with 1. Here’s the output:
installer: Package name is Oracle VM VirtualBox
installer: Installing at base path /```

You will need to install the virtual box cast and make a change in Security & Privacy

In a terminal run:

`brew cask install virtualbox`

Then Open up System Preferences > Security & Privacy

you’ll see a message saying System software from developer “Oracle America, Inc.” was blocked from loading. If you do see the message, click the lock icon in the bottom left of the window Enter your Mac password Then click the Allow button.

Rerun `curl -s https://raw.githubusercontent.com/classyllama/devenv/master/vagrant/bin/install.sh  https://raw.githubusercontent.com/classyllama/devenv/master/vagrant/bin/devbox.sh | bash /dev/stdin --lib-mode` 

## Development Notes

### Session Storage
It is well recognized that PHP cannot store sessions on an NFS mount. Since `/var/www/sites/` is mounted in the vm via an NFS mount, this causes trouble with storing session files inside the document root. Magento 2 seems to handle this just fine and stores it's sessions in the configured session location. Magento 1 requires a workaround to function.

To workaround this issue, replace the `var/session` directory with a soft-link pointing at the default php session store:

```bash
rm -rf var/session
ln -s /var/lib/php/session var/session
```

Alternately, you may use an alternative session storage mechanism such as redis or memcached to store sessions and avoid the problem altogether.

## VMWare Provider

Using VMWare Fusion is a supported (but non-default) setup. There are additional steps involved to use it due to differences in how Virtual Box and VMX configure network interfaces and handle NFS mounts.

For NFS mounts to function, run the following to add the necessary exports to your `/etc/exports` file on the host machine and restart nfsd:

```bash
MAPALL="-mapall=$(id -u):$(grep ^admin: /etc/group | cut -d : -f 3)"
MOUNT_DIR="$(readlink /server || echo /server)"
printf "%s\n%s\n" \
    "$MOUNT_DIR/sites/ -alldirs -network 192.168.235.0 -mask 255.255.255.0 $MAPALL" \
    "$MOUNT_DIR/mysql/ -alldirs -network 192.168.235.0 -mask 255.255.255.0 $MAPALL" \
    | sudo tee -a /etc/exports > /dev/null
sudo nfsd restart
```

# License
This project is licensed under the Open Software License 3.0 (OSL-3.0). See included LICENSE file for full text of OSL-3.0


CHANGELOG.md
2.0.21
===============

* web73 node

When you run vagrant for the first time after updating it will likely prompt you for credentials so that it can add an entry to your /etc/hosts file for the new web73 machine.

Some PHP Extensions do not exists on the web73 instance due to them not being available in the IUS repo

  mhash
  curl
  xsl
  zip

2.0.20
===============

* Addition of PHP libsodium support on web72 instance for Magento 2.3.2 requirements

2.0.19
===============

* Install magento-cloud CLI tool

2.0.18
===============

* Updated base image to bento/centos-7 (previously bento/centos-7.5)
  bento/centos-7: CentOS Linux release 7.6.1810
* Changed ansible roles to use classyllama roles
* Updates to classyllama/repo_ius adds old IUS GPG key to fix validating older packages from ius-archive

2.0.17
===============

* Worked around issue with IUS Repo GPG Key Check validations
  Disabled GPG Check on all IUS repos
  Issues were seen after ius-release was updated to version 2-1.el7.ius from 1.0-15.ius.centos7

2.0.16
===============

* Fixed issue where redis32u was not found when spinning up servers as it was moved to the ius archive repo

2.0.14
===============

* Fixed issue where php56u was not found when spinning up web56 as it was moved to the ius archive repo

2.0.13
===============

* improved mysql config so that magento 2.3.0 would finish install and not say "MySQL server has gone away"

2.0.12
===============

* update percona repo

2.0.11
===============

* Fixed issue where php70u was not found when spinning up web70 as it was moved to the ius archive repo

2.0.10
===============

* Fixed issue where vagrant 2.2.1 spins up VMs with the NIC type "virtio" which breaks the NFS mounts
* Fixed a few asserts in install.sh and devbox.sh which broke the install process on a new host machine

2.0.9
===============

* Fixed issues in dependencies which resulted in Ansible 2.7 stable breaking box builds (PR #108)
* Changed `varnish-dev` service to `varnish` services (no longer install varnish as a "named" instance)
* Changed dependencies for multi-redis and php-fpm services to use public ansible galaxy roles

2.0.8
===============

* Changed default box to web71 and disabled auto-start on web70
* Updated base image to bento/centos-7.5 (previously bento/centos-7.3)
* Updated the following roles to pull in various improvements from upstream for nginx, varnish, and redis

2.0.7
===============

* Added web72 box with PHP 7.2
* Updated install.sh to install PHP on macOS host via new brew packages vs relying on the defunct php taps

2.0.6
===============

* Added web based dev-web70/opcache.php file for inspecting the mod_php opcache statistics (for real this time!)

2.0.5
===============

* Added CLI based "cachetool" for inspecting the php-fpm opcache statistics
* Added web based dev-web70/opcache.php file for inspecting the mod_php opcache statistics
* Updated alpacaglue.php-fpm role to pull in php-opcache optimization settings
* Updated version check in m2setup.sh to play nicely with the Magento 2.3.0 alpha release

2.0.4
===============

* Added an `ansible_roles.yml` file declaring what roles are pulled in from an external source (some of them come from a private upstream, but this lines out which may be modified directly in this repository).
* Added explicit declaration of connection type (using `local`) on ansible provisiner making running the following inside the VM later on possible (handy for developing changes to provisioners):

        ansible-playbook -ilocalhost, /vagrant/provisioning/<playbook_name>.yml

* Added `compatibility_mode` value to ansible provisioner declaration to quite some warnings on VM build
* Fixed issue in elasticsearch role causing VM build to fail due to role's incompatiblity with ansible 2.5 or newer
* Updated all roles to use correct namespace prefixes to identify where they came from
* Updated all roles pulling in all external role improvements to-date

2.0.3
===============

* Reverted workaround for bad autoloader failing m2setup.sh when using the --sampledata flag
* Updated composer version to latest v1.6.3 (resolves underlying bug failing sample data deployment)

2.0.2
===============

* Added support for vim with plugins located at /server/.shared/vim/plugin/*.vim
* Added workaround for bad autoloader failing m2setup.sh when using the --sampledata flag

2.0.1
===============

* Added `patch` to list of utils installed by default in each vm
* Fixed issue causing a `halt` to forcefully shutdown a VM (raised `graceful_halt_timeout` in config)
* Fixed issue causing `install.sh` to fail on a from-scratch setup
* Updated redis provisioning role to install from ius-archive (versions prior to 4.0 no longer available in "stable" rpm)
* Updated installed version of composer to 1.6.1 (previously was 1.1.0)
* Updated m2setup.sh to require use of explicit --branch flag when using -g / --github
* Removed auto-configuration of ~/.my.cnf file due to lack of central db node in v2.0
* Removed `develop` as valid option for --branch flag in m2setup.sh script
* Disabled clipmenu brew package in devbox.sh install script additions file

2.0.0
===============

**NOTE: To upgrade from a prior version to version 2.0.0 or newer, please reference the procedure outlined in UPGRADE.md. There are many core changes that must be accounted for. I.e. a simple destroy/up of the VM will not be sufficient in this case!**

* Added display of current terraform workspace to PS1 when used
* Added web71 machine
* Added Solr provisioning role
* Added elasticsearch provisioning role
* Added local cache of Solr install files
* Added public to web-dirs vhosts.sh will use when configuring nginx
* Added support to vcd for specifying a node name it will ssh into
* Fixed issue with m2setup.sh failing to install sample-data in Magento 2.1 composer installs due to bin/magento change
* Fixed issue in install.sh causing a failure due to missing brew cask package

vagrant/vagrant.rb
##
 # Copyright © 2015 by David Alger. All rights reserved
 # 
 # Licensed under the Open Software License 3.0 (OSL-3.0)
 # See included LICENSE file for full text of OSL-3.0
 # 
 # http://davidalger.com/contact/
 ##

require_relative 'lib/config'
require_relative 'lib/mount'
require_relative 'lib/provision'
require_relative 'lib/machine'

# configure environment paths
BASE_DIR = base_dir('/Volumes/Server')
MOUNT_PATH = mount_path(BASE_DIR)
VAGRANT_DIR = BASE_DIR + '/vagrant'
SHARED_DIR = BASE_DIR + '/.shared'
SITES_DIR = '/sites'
SITES_MOUNT = '/var/www/sites'

# auto configure host machine
auto_config_host

# begin the configuration sequence
Vagrant.require_version '>= 1.9.5'
Vagrant.configure(2) do |conf|
  conf.vm.define :web70, autostart: false do |node|
    configure_basebox node, host: 'dev-web70', ip: '10.19.89.14'
    configure_web node
    configure_percona node
    configure_solr node
    configure_elasticsearch node
  end

  conf.vm.define :web71, primary: true do |node|
    configure_basebox node, host: 'dev-web71', ip: '10.19.89.15'
    configure_web node, php_version: 71
    configure_percona node
    configure_solr node
    configure_elasticsearch node
  end

  conf.vm.define :web72, autostart: false do |node|
    configure_basebox node, host: 'dev-web72', ip: '10.19.89.16'
    configure_web node, php_version: 72
    configure_percona node
    configure_solr node
    configure_elasticsearch node
  end

  conf.vm.define :web56, autostart: false do |node|
    configure_basebox node, host: 'dev-web56', ip: '10.19.89.10'
    configure_web node, php_version: 56
    configure_percona node
    configure_solr node
    configure_elasticsearch node
  end

  conf.vm.define :web55, autostart: false do |node|
    configure_basebox node, host: 'dev-web55', ip: '10.19.89.11'
    configure_web node, php_version: 55
    configure_percona node
    configure_solr node
    configure_elasticsearch node
  end
end


vagrant/lib/mount.rb
##
 # Copyright © 2015 by David Alger. All rights reserved
 # 
 # Licensed under the Open Software License 3.0 (OSL-3.0)
 # See included LICENSE file for full text of OSL-3.0
 # 
 # http://davidalger.com/contact/
 ##

class Mount
  @@mounts = []

  # Mounts a directory from the host machine on the vm at the designated path using an NFS mount
  # Params:
  # +host_path+:: +String+ full path to directory on host machine
  # +guest_path+:: +String+ mount location on guest machine
  # +mount_options+:: +Array+ additional mount options
  def self.nfs (id, host_path, guest_path, mount_options: [])
    @@mounts += [{ type: 'nfs', id: id, host_path: host_path, guest_path: guest_path, mount_options: mount_options }]
  end

  # Mounts directory from the host machine on the vm at the designated path using default vmfs
  # Params:
  # +host_path+:: +String+ full path to directory on host machine
  # +guest_path+:: +String+ mount location on guest machine
  # +mount_options+:: +Array+ additional mount options
  def self.vmfs (id, host_path, guest_path, mount_options: [])
    @@mounts += [{ type: 'vmfs', id: id, host_path: host_path, guest_path: guest_path, mount_options: mount_options }]
  end

  # Binds location on guest machine from provided source to target
  # Params:
  # +source_path+:: +String+ full path to directory on host machine
  # +target_path+:: +String+ mount location on guest machine
  def self.bind (source_path, target_path)
    @@mounts += [{ type: 'bind', source_path: source_path, target_path: target_path }]
  end

  # Asserts that an entry is in the /etc/exports file to gaurantee that an NFS mount is possible
  # Params:
  # +host_path+:: +String+ path to required share directory
  def self.assert_export (host_path)
    if File.exist?('/etc/exports') == false
      $stderr.puts "Error: /etc/exports does not exist. See /Volumes/Server/README.md for details"
      exit false
    end

    exports = File.readlines('/etc/exports')
    for line in exports
      if line.start_with?(host_path + '/ -alldirs')
        return true
      end
    end
    $stderr.puts "Error: /etc/exports is missing an entry for #{host_path}/. See /Volumes/Server/README.md for details"
    exit false
  end

  # Sets up the vagrant configuration neccesary for the mounts configured via the Mount class
  # Params:
  # +conf+:: vagrant provisioning conf object
  def self.provision (conf)
    bindings = []

    @@mounts.each do |mount|
      case mount[:type]
        when 'nfs'
          provision_nfs(conf, mount)
        when 'vmfs'
          provision_vmfs(conf, mount)
        when 'bind'
          bindings += [mount]
      end
    end

    if bindings.count
      provision_bind(conf, bindings)
    end

    # reset so subsequent calls won't operate on the same mounts as a previous run
    @@mounts = []
    bindings = []
  end

  def self.provision_nfs (conf, mount)
    conf.vm.synced_folder mount[:host_path], mount[:guest_path], id: mount[:id], create: true, nfs_export: false,
      type: 'nfs', :nfs => { mount_options: mount[:mount_options] }
  end
  private_class_method :provision_nfs

  def self.provision_vmfs (conf, mount)
    if mount[:id].start_with?('-')
      throw "Error: mount_vmfs id may not begin with a hyphen, id of #{mount[:id]} given"
    end

    FileUtils.mkdir_p mount[:host_path]   # ensure path on host is present
    conf.vm.synced_folder mount[:host_path], mount[:guest_path], id: mount[:id], mount_options: mount[:mount_options]
  end
  private_class_method :provision_vmfs

  def self.provision_bind (conf, bind)
    bind_sh = ""

    bind.each do | mount |
      bind_sh = %-#{bind_sh}
        echo "#{mount[:source_path]} =\> #{mount[:target_path]}"
        mkdir \-p "#{mount[:source_path]}" "#{mount[:target_path]}"
        sudo mount \-o bind "#{mount[:source_path]}" "#{mount[:target_path]}"
      -
    end

    conf.vm.provision :shell, run: 'always' do |conf|
      conf.name = "bind_sh"
      conf.inline = bind_sh
    end
  end
  private_class_method :provision_bind
end


vagrant/lib/config.rb
##
 # Copyright © 2015 by David Alger. All rights reserved
 # 
 # Licensed under the Open Software License 3.0 (OSL-3.0)
 # See included LICENSE file for full text of OSL-3.0
 # 
 # http://davidalger.com/contact/
 ##

# Validates path as a valid symlink pointing to our env or the containing directory of our env. Returns path
# Params:
# +base_dir+:: +String+ path to base directory or symlink to environment root
def base_dir (base_dir)
  # evaluates to true if is not either a valid directory a symlink pointing to a directory
  if !File.directory?(base_dir)
    throw 'Error: please create a /Volumes/Server link pointing to the environment root'
  end

  # assert base_dir points to our environment root
  if %-#{mount_path(base_dir)}/#{File.basename(VAGRANT_FILE)}- != VAGRANT_FILE
    throw "Error: #{base_dir} does not point at this environment root (did you link with a trailing slash?)"
  end
  
  return base_dir
end

# Resolves the mount path from the base_dir. Will return unchanged if base_dir is not a symlink.
# Params:
# +base_dir+:: +String+ path to base directory
def mount_path (base_dir)
  if File.symlink?(base_dir)
    base_dir = File.readlink(base_dir)
  end
  
  if !File.directory?(base_dir)
    throw "Error: #{base_dir} does not exist"
  end
  
  return base_dir
end

# Asserts that an entry is in the /etc/hosts file for the given host / ip combination
# Params:
# +host+:: +String+ hostname of the hosts file entry
# +ip+:: +String+ ip address of the hosts file entry
def assert_hosts_entry (host, ip)
  if not %x{grep -E '^#{ip}\\W+#{host}$' /etc/hosts}.strip!
    puts "==> host: Appending '#{ip} #{host}' to /etc/hosts file"
    system %-echo '#{ip} #{host}' | sudo tee \-a /etc/hosts > /dev/null-
  end
end

# Runs the host machine autoconfiguration routine
# Params:
def auto_config_host
  changes = false
  newsh = false

  assert_hosts_entry 'dev-host', '10.19.89.1'

  # place flag on host machine for use in common shell scripts
  if not File.exist?('/etc/.vagranthost')
    puts '==> host: Touching host indicator'
    system %-sudo touch /etc/.vagranthost-
    changes = true
  end

  # create symlink to profile.d scripts
  if not File.symlink?('/etc/profile.d')
    puts "==> host: Linking /etc/profile.d -> #{VAGRANT_DIR}/etc/profile.d"
    system %-sudo ln \-s #{VAGRANT_DIR}/etc/profile.d /etc/profile.d-
    changes = true
    newsh = true
  end

  # enable profile.d scripts on host
  if not %x{grep '## VAGRANT START ##' /etc/profile}.strip!
    puts "==> host: Configuring /etc/profile for running sub-scripts"

    profile_script=%-
for i in /etc/profile.d/*.sh ; do
    if [ \-r "$i" ]; then
        if [ "${\-#*i}" != "$\-" ]; then
            . "$i"
        else
            . "$i" >/dev/null 2>&1
        fi
    fi
done
unset i
-

    # append script to /etc/profile and ~/.zshrc and then execute that new portion to pickup env configuration
    system %-
      printf "\n## VAGRANT START ##%s## VAGRANT END ##\n" '#{profile_script}'| sudo tee \-a /etc/profile > /dev/null
    -
    system %-
      printf "\n## VAGRANT START ##%s## VAGRANT END ##\n" '#{profile_script}'| sudo tee \-a ~/.zshrc > /dev/null
    -
    changes = true
    newsh = true
  end
  
  # add exports for NFS mounts to /etc/exports
  if not (File.exist?('/etc/exports') and %x{grep '## VAGRANT START ##' /etc/exports}.strip!)
    puts "==> host: Adding entries to /etc/exports for NFS mounts"
    
    mapall = %x{printf $(id \-u):$(grep ^admin: /etc/group | cut \-d : \-f 3)}
    nfs_exports = %-
#{MOUNT_PATH}/sites/ \-alldirs \-network 10.19.89.0 \-mask 255.255.255.0 \-mapall=#{mapall}
#{MOUNT_PATH}/mysql/ \-alldirs \-network 10.19.89.0 \-mask 255.255.255.0 \-mapall=#{mapall}
-
    system %-
      printf "\n## VAGRANT START ##%s## VAGRANT END ##\n" '#{nfs_exports}' | sudo tee \-a /etc/exports > /dev/null
      sudo nfsd restart
    -
    changes = true
  end

  # verify virtualbox machine directory
  vbox_machine_dir = %x{VBoxManage list systemproperties | grep 'Default machine folder:' | sed 's/.*: *//g'}.strip!
  if vbox_machine_dir != "#{BASE_DIR}/.machines"
    puts "==> host: Setting global VirtualBox machine folder to /Volumes/Server/.machines"
    system %-VBoxManage setproperty machinefolder #{BASE_DIR}/.machines-
    changes = true
  end

  if changes
    puts '==> host: Auto configuration complete'
  end

  # prevent run if shell doesn't have expected env vars set from profile.d scripts or if changes require such
  if newsh or (changes == false && ENV['VAGRANT_IS_SETUP'] != 'true')
    puts 'Please re-run the command in a new shell... or type `source /etc/profile` (bash) or `source ~/.zshrc` (zsh) and then try again'
    exit 1
  end
end


vagrant/lib/machine.rb
##
 # Copyright © 2015 by David Alger. All rights reserved
 #
 # Licensed under the Open Software License 3.0 (OSL-3.0)
 # See included LICENSE file for full text of OSL-3.0
 #
 # http://davidalger.com/contact/
 ##

def configure_basebox (node, host: nil, ip: nil, memory: 4096, cpu: 2)
  node.vm.box = 'bento/centos-7'

  node.vm.hostname = host
  node.vm.network :private_network, ip: ip, nic_type: "82540EM"
  node.vm.provider "virtualbox" do |vb|
    vb.customize ["modifyvm", :id, "--nictype1", "82540EM"]
  end
  assert_hosts_entry host, ip

  node.vm.graceful_halt_timeout = 120

  node.vm.synced_folder VAGRANT_DIR, '/vagrant'

  # mount persistent shared cache storage on vm and bind sub-caches
  Mount.vmfs('host-cache', SHARED_DIR, SHARED_DIR)
  Mount.bind(SHARED_DIR + '/yum', '/var/cache/yum')
  Mount.bind(SHARED_DIR + '/npm', '/var/cache/npm')

  # setup mount provisioners
  Mount.provision(node)

  # setup the yum cache before vagrant installs ansible in the box (so it can be installed from the cache)
  node.vm.provision :shell do |conf|
    conf.name = "yum-cache"
    conf.inline = 'sed -i \'s/keepcache=0/keepcache = 1\nmetadata_expire = 24h/\' /etc/yum.conf'
  end

  # run the basebox ansible play on the box
  ansible_play(node, 'basebox', {
    host_zoneinfo: File.readlink('/etc/localtime')
  })

  # configure default RAM and number of CPUs allocated to vm
  vm_set_ram(node, memory)
  vm_set_cpu(node, cpu)

  # so we can connect to remote servers from inside the vm
  node.ssh.forward_agent = true
end

def configure_web (node, php_version: 70)
  # verify exports and mount nfs sites location
  Mount.assert_export(MOUNT_PATH + SITES_DIR)
  Mount.nfs('host-www-sites', MOUNT_PATH + SITES_DIR, SITES_MOUNT)

  # bind localhost pub directory
  #Mount.bind(SITES_MOUNT + '/__localhost/pub', '/var/www/html')
  Mount.vmfs('host-html', MOUNT_PATH + SITES_DIR + '/__localhost/pub', '/var/www/html')

  # setup guest provisioners
  Mount.provision(node)
  ansible_play(node, 'web', {
    php_version: php_version,
    shared_ssl_dir: SHARED_DIR + '/ssl'
  })

  # run vhosts.sh on every reload
  node.vm.provision :shell, run: 'always' do |conf|
    conf.name = "vhosts.sh"
    conf.inline = "vhosts.sh --quiet"
  end
end

def configure_percona (node, mysql_version: 56)
  # verify exports and mount nfs mysql data directory
  Mount.assert_export(MOUNT_PATH + '/mysql')
  Mount.nfs('host-mysql-data', MOUNT_PATH + '/mysql/' + node.vm.hostname.sub('dev-', ''), '/var/lib/mysql')

  # setup guest provisioners
  Mount.provision(node)
  ansible_play(node, 'percona', {
    mysql_version: mysql_version
  })

  # start mysqld on every reload (must happen here so mysqld starts after file-system is mounted)
  node.vm.provision :shell, run: 'always' do |conf|
    conf.name = "systemctl start mysqld"
    conf.inline = "systemctl start mysqld"
  end
end

def configure_elasticsearch (node)
  ansible_play(node, 'elasticsearch')
end

def configure_solr (node)
  # bind the solr workspace to the install/download cache location
  Mount.bind(SHARED_DIR + '/solr', '/var/cache/solr')
  Mount.provision(node)

  # provision the box with solr
  ansible_play(node, 'solr')
end
